### PR TITLE
[camerax] Adds functionality to bind UseCases to a lifecycle

### PR DESCRIPTION
**Pull request by camsim99**
Adds `UseCase` and `Camera` classes and adds methods in the `ProcessCameraProvider` class to allow for the binding of `UseCase`s to the lifecycle of a `Camera`.

Fixes https://github.com/flutter/flutter/issues/111220.
Part of https://github.com/flutter/flutter/issues/111127, https://github.com/flutter/flutter/issues/115846.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraAndroidCameraxPlugin.java [hunk 1]
```diff
@@ -66,7 +63,11 @@ public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
 
   @Override
   public void onAttachedToActivity(@NonNull ActivityPluginBinding activityPluginBinding) {
-    updateContext(activityPluginBinding.getActivity());
+    CameraAndroidCameraxPlugin plugin = new CameraAndroidCameraxPlugin();
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraAndroidCameraxPlugin.java:66]: I think I missed this in the last PR. But I don't think it is necessary to create a new `CameraAndroidCameraxPlugin` instance. You should be able to just call `setUp` on the current instance, right?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java [hunk 2]
```diff
@@ -84,4 +91,52 @@ public List<Long> getAvailableCameraInfos(@NonNull Long identifier) {
...
+   * corresponds to this instance and returns the instance of the {@code Camera} whose lifecycle
+   * that {@code LifecycleOwner} reflects.
+   */
+  @Override
+  public Long bindToLifecycle(
+      @NonNull Long identifier,
+      @NonNull Long cameraSelectorIdentifier,
+      @NonNull List<Long> useCaseIds) {
+    ProcessCameraProvider processCameraProvider =
+        (ProcessCameraProvider) instanceManager.getInstance(identifier);
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java:106]: Could you wrap `instanceManager.getInstance(..)` with `Objects.requireNonnull(...)`. It helps with debugging and removes the lint warning. Same for the line below and other apis.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraTest.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraTest.java:24]: Does this test need to be run with `RobolectricTestRunner`? I noticed the other tests have this as well. So if no, then they can remove it as well.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  @Rule public MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  @Mock public BinaryMessenger mockBinaryMessenger;
+  @Mock public Camera camera;
+
+  InstanceManager testInstanceManager;
+
+  @Before
+  public void setUp() {
+    testInstanceManager = InstanceManager.open(identifier -> {});
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/CameraTest.java:35]: This test is missing 

```java
@After
  public void tearDown() {
    testInstanceManager.close();
  }
```

File: packages/camera/camera_android_camerax/lib/src/process_camera_provider.dart [hunk 5]
```diff
@@ -69,22 +91,73 @@ class ProcessCameraProviderHostApiImpl extends ProcessCameraProviderHostApi {
...
   }
 
-  /// Retrives the list of CameraInfos corresponding to the available cameras.
-  Future<List<CameraInfo>> getAvailableCameraInfosFromInstances(
-      ProcessCameraProvider instance) async {
+  /// Gets identifier that the [instanceManager] has set for
+  /// the [ProcessCameraProvider] instance.
+  int getProcessCameraProviderIdentifier(ProcessCameraProvider instance) {
     int? identifier = instanceManager.getIdentifier(instance);
     identifier ??= instanceManager.addDartCreatedInstance(instance,
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/lib/src/process_camera_provider.dart:98]: Why does this add the instance to the `InstanceManager` if it doesn't contain the `ProcessCameraProvider`.?Shouldn't `getInstancefromInstances` add the `ProcessCameraProvider` to the `InstanceManager`? This should throw an error if the `InstanceManager` doesn't contain `instance`.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java (hunk 2 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java:106]: cc @gmackall just a heads up about this

File: packages/camera/camera_android_camerax/lib/src/process_camera_provider.dart (hunk 5 above)
Comment (camsim99) [reply packages/camera/camera_android_camerax/lib/src/process_camera_provider.dart]: I agree. Fixed that.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java [hunk 6]
```diff
@@ -84,4 +92,59 @@ public List<Long> getAvailableCameraInfos(@NonNull Long identifier) {
...
+                  instanceManager.getInstance(((Number) useCaseIds.get(i)).longValue()));
+    }
+
+    Camera camera =
+        processCameraProvider.bindToLifecycle(
+            (LifecycleOwner) lifecycleOwner, cameraSelector, useCases);
+
+    final CameraFlutterApiImpl camraFlutterApi =
+        new CameraFlutterApiImpl(binaryMessenger, instanceManager);
+    camraFlutterApi.create(camera, result -> {});
```
Comment (bparrishMines) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/ProcessCameraProviderHostApiImpl.java:125]: nit 
```suggestion
    final CameraFlutterApiImpl cameraFlutterApi =
        new CameraFlutterApiImpl(binaryMessenger, instanceManager);
    cameraFlutterApi.create(camera, result -> {});
```

**Review by bparrishMines — APPROVED** [review]
LGTM with nits

**Pull request merged**